cpp/src/quantiles/quantiles.cu

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+#include <cudf/copying.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/types.hpp>
+#include <quantiles/quantiles_util.hpp>
+#include "cudf/utilities/traits.hpp"
+#include "thrust/functional.h"
+#include "thrust/iterator/counting_iterator.h"
+
+using ScalarResult = double;
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+
+// enum class extrema {
+//     min,
+//     max
+// };
+
+// template<extrema minmax, bool is_nullable>
+// size_type extrema(column_view const & in,
+//                   cudf::order order,
+//                   cudf::null_order null_order,
+//                   cudaStream_t stream)
+// {
+//     std::vector<cudf::order> h_order{ order };
+//     std::vector<cudf::null_order> h_null_order{ null_order };
+//     rmm::device_vector<cudf::order> d_order( h_order );
+//     rmm::device_vector<cudf::null_order> d_null_order( h_null_order );
+//     table_view in_table({ in });
+//     auto in_table_d = table_device_view::create(in_table);
+//     auto it = thrust::make_counting_iterator<size_type>(0);
+
+//     auto comparator = row_lexicographic_comparator<is_nullable>(
+//         *in_table_d,
+//         *in_table_d,
+//         d_order.data().get(),
+//         d_null_order.data().get());
+
+//     auto extrema_id = minmax == extrema::min
+//         ? thrust::min_element(rmm::exec_policy(stream)->on(stream), it, it + in.size(), comparator)
+//         : thrust::max_element(rmm::exec_policy(stream)->on(stream), it, it + in.size(), comparator);
+
+//     return *extrema_id;
+// }
+
+template<typename T>
+std::unique_ptr<scalar>
+pick(column_view const& in, size_type index) {
+    auto result = get_array_value<ScalarResult>(in.begin<T>(), index);
+    return std::make_unique<numeric_scalar<ScalarResult>>(result);
+}
+
+struct pick_functor
```

Comment:
I was using this to obtain values for the `q < 0 || q > 1` case, but that's commented out for the moment.